cpp/src/arrow/filesystem/filesystem_utils.h

```diff
@@ -0,0 +1,109 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_FILESYSTEM_FACTORY_H
+#define ARROW_FILESYSTEM_FACTORY_H
+
+#include <memory>
+#include <string>
+#include "arrow/filesystem/filesystem.h"
+
+namespace arrow {
+
+namespace fs {
+
+enum class FileSystemType { HDFS, LOCAL, S3, UNKNOWN };
+
+/// \class PathInfo
+/// \brief A class used to parse URI-based path.
+class PathInfo {
```

Comment:
I made PathInfo public is because after we getting the std::shared_ptr<FileSystem> instance, we will also need to use path_info->GetFileName() and path_info-> GetDirectoryName() to see what file is asked in the URI-based path, take [filesystem_utils_test#133](https://github.com/apache/arrow/pull/5717/files#diff-264640773a55e85c8a1c7247535c7ec3R133) as an example.

Reply:
It seems that only the filesystem factory and unit tests will need to make that query, so PathInfo can be moved to an internal header.

Reply:
@bkietz, oh, I aimed to use unit tests to indicate that later on who used file systemfactory will need Pathinfo
If that doesn’t make clear to you, do you mind to take a look of my following pr #5719 ? It will need pathinfo to open a parquet input file and output stream [check this link to line of codes](https://github.com/apache/arrow/pull/5719/commits/d9ec1e42d820970dbf04b3f05ac1731fec04d617#diff-cf191b4f5bc165e351880778db1c5ef1R102)

Reply:
@bkietz , as I explained before, I had submitted a huge pr #5522 before and @emkornfield asked me to make the pr into smaller ones , so I submitted two PR #5717 and #5719 for file system unified api and parquet jni bridge, that is why I needed this pathinfo class as a public one, hope this makes sense to you!

Reply:
Your only usage of it outside file system factory in that PR is [extracting the directory and file name in the jni bridge](https://github.com/apache/arrow/pull/5719/files#diff-cf191b4f5bc165e351880778db1c5ef1R281). If you need that to be public then just expose (for example) `std::string HadoopFileSystem::GetPath(const std::string& uri)` as a separate helper function and make `PathInfo` private

Reply:
hi, @bkietz , I am not sure if I get what you proposed, I thought hadoopfilesystem is a class for filesystem operations, such as create dir, delete dir, open read file and open write file, so we should not store infomations like filepath inside hadoopfilesystem, am I right?
And if I expose a function called getFilePath inside hadoopfilesystem, then I still need to expose one inside s3fs and localfs when I need to open files through path, why don't we just put this as an public PathInfo class inside filesystem_utils.h? 
Do you mind if we can discuss on this in an IM?  I felt like we can only discuss on this one time a day, which make the conversion very slow. my gmail account is xuechendi@gmail.com

Reply:
That's not possible, but you could post to the mailing list dev@apache.arrow.org

Reply:
If the only use case for the PathInfo class is to retrieve file paths from uris then that doesn't warrant a public class. You're right that it applies to more than HDFS, maybe you could make it a static member of Uri:
```c++
Result<std::string> Uri::GetPath(const std::string& uri) {
  Uri parsed;
  RETURN_NOT_OK(parsed.Parse(uri));
  return parsed.path();
}
```

Reply:
@bkietz , I see, I’ll add it to URI, thanks